File: days/day_04/solutions/solution.clj
(require '[clojure.string :as str])

;; Import utilities (assuming they're on classpath)
;; For now, we'll use inline file reading for simplicity
(defn read-input [path]
  (->> path
       slurp
       str/trim
       str/split-lines))

(defn count-neighbors [grid i j rows cols]
  (reduce (fn [cnt [di dj]]
            (if (and (= di 0) (= dj 0))
              cnt
              (let [ni (+ i di)
                    nj (+ j dj)]
                (if (and (>= ni 0) (< ni rows) (>= nj 0) (< nj cols) 
                         (= (get-in grid [ni nj]) \@))
                  (inc cnt)
                  cnt))))
          0
          (for [di [-1 0 1] dj [-1 0 1]] [di dj])))

(defn solve [lines]
  (let [rows (count lines)
        cols (if (> rows 0) (count (first lines)) 0)
        
        ;; Part 1: Count accessible rolls (fewer than 4 neighbors that are '@')
        part1-count (reduce (fn [cnt [i j]]
                              (if (= (get-in lines [i j]) \@)
                                (let [neighbors (count-neighbors lines i j rows cols)]
                                  (if (< neighbors 4)
                                    (inc cnt)
                                    cnt))
                                cnt))
                            0
                            (for [i (range rows) j (range cols)] [i j]))
        
        ;; Part 2: Iteratively remove accessible rolls until none can be removed
        grid (atom (mapv vec lines))
        part2-count (loop [count 0]
                      (let [to-remove (reduce (fn [acc [i j]]
                                                (if (= (get-in @grid [i j]) \@)
                                                  (let [neighbors (count-neighbors @grid i j rows cols)]
                                                    (if (< neighbors 4)
                                                      (conj acc [i j])
                                                      acc))
                                                  acc))
                                              []
                                              (for [i (range rows) j (range cols)] [i j]))]
                        (if (empty? to-remove)
                          count
                          (do
                            (doseq [[i j] to-remove]
                              (swap! grid assoc-in [i j] \.))
                            (recur (+ count (count to-remove)))))))]
    [part1-count part2-count]))

(let [lines (read-input "../data/input.txt")
      [part1 part2] (solve lines)]
  (println (str "Part 1: " part1))
  (println (str "Part 2: " part2)))

File: days/day_04/solutions/solution.ex
defmodule Solution do
  def count_neighbors(grid, i, j, rows, cols) do
    Enum.sum(for di <- -1..1, dj <- -1..1, not (di == 0 and dj == 0) do
      ni = i + di
      nj = j + dj
      if ni >= 0 and ni < rows and nj >= 0 and nj < cols and
         Enum.at(Enum.at(grid, ni), nj) == "@" do
        1
      else
        0
      end
    end)
  end

  def solve(lines) do
    rows = length(lines)
    cols = if rows > 0, do: String.length(hd(lines)), else: 0

    # Part 1: Count accessible rolls (fewer than 4 neighbors that are '@')
    part1 = Enum.sum(for i <- 0..(rows-1), j <- 0..(cols-1) do
      line = Enum.at(lines, i)
      if String.at(line, j) == "@" do
        neighbors = count_neighbors(lines, i, j, rows, cols)
        if neighbors < 4, do: 1, else: 0
      else
        0
      end
    end)

    # Part 2: Iteratively remove accessible rolls until none can be removed
    grid = Enum.map(lines, &String.graphemes/1) |> Enum.map(&List.to_tuple/1)
    {part2, _} = remove_iteratively(grid, rows, cols, 0)

    {part1, part2}
  end

  defp remove_iteratively(grid, rows, cols, count) do
    to_remove = for i <- 0..(rows-1), j <- 0..(cols-1), into: [] do
      if elem(elem(grid, i), j) == "@" do
        grid_list = Enum.map(0..(rows-1), fn r ->
          Tuple.to_list(elem(grid, r))
        end)
        neighbors = count_neighbors(grid_list, i, j, rows, cols)
        if neighbors < 4, do: {i, j}, else: nil
      else
        nil
      end
    end |> Enum.filter(&(&1 != nil))

    if Enum.empty?(to_remove) do
      {count, grid}
    else
      new_grid = Enum.reduce(to_remove, grid, fn {i, j}, g ->
        row = elem(g, i)
        new_row = put_elem(row, j, ".")
        put_elem(g, i, new_row)
      end)
      remove_iteratively(new_grid, rows, cols, count + length(to_remove))
    end
  end
end

lines = "../data/input.txt" |> File.read!() |> String.trim() |> String.split("\n")
{part1, part2} = Solution.solve(lines)
IO.puts("Part 1: #{part1}")
IO.puts("Part 2: #{part2}")
